example.js

```diff
@@ -0,0 +1,151 @@
+'use strict'
+
+const Fastify = require('fastify')
+
+function build (opts) {
+  const fastify = Fastify(opts)
+
+  fastify
+    .register(require('fastify-jwt'), { secret: 'supersecret' })
+    .register(require('fastify-leveldb'), { name: 'authdb' })
+    .register(require('./fastify-auth'))
+    .after(routes)
+
+  fastify.decorate('verifyJWTandLevel', verifyJWTandLevel)
+  fastify.decorate('verifyUserAndPassword', verifyUserAndPassword)
+
+  function verifyJWTandLevel (request, reply, done) {
+    const jwt = this.jwt
+    const level = this.level
+
+    if (!request.req.headers['auth']) {
+      return done(new Error('Missing token header'))
+    }
+
+    jwt.verify(request.req.headers['auth'], onVerify)
+
+    function onVerify (err, decoded) {
+      if (err || !decoded.user || !decoded.password) {
+        return done(new Error('Token not valid'))
+      }
+
+      level.get(decoded.user, onUser)
+
+      function onUser (err, password) {
+        if (err) {
+          if (err.notFound) {
+            return done(new Error('Token not valid'))
+          }
+          return done(err)
+        }
+
+        if (!password || password !== decoded.password) {
+          return done(new Error('Token not valid'))
+        }
+
+        done()
+      }
+    }
+  }
+
+  function verifyUserAndPassword (request, reply, done) {
+    const level = this.level
+
+    level.get(request.body.user, onUser)
+
+    function onUser (err, password) {
+      if (err) {
+        if (err.notFound) {
+          return done(new Error('Password not valid'))
+        }
+        return done(err)
+      }
+
+      if (!password || password !== request.body.password) {
+        return done(new Error('Password not valid'))
+      }
+
+      done()
+    }
+  }
+
+  function routes () {
+    fastify.route({
+      method: 'POST',
+      url: '/register',
+      schema: {
+        body: {
+          type: 'object',
+          propertries: {
+            user: { type: 'string' },
+            password: { type: 'string' }
+          },
+          required: ['user', 'password']
+        }
+      },
+      handler: (req, reply) => {
+        req.log.info('Creating new user')
+        fastify.level.put(req.body.user, req.body.password, onPut)
+
+        function onPut (err) {
+          if (err) return reply.send(err)
+          fastify.jwt.sign(req.body, onToken)
+        }
+
+        function onToken (err, token) {
+          if (err) return reply.send(err)
+          req.log.info('User created')
+          reply.send({ token })
+        }
+      }
+    })
+
+    fastify.route({
+      method: 'GET',
+      url: '/no-auth',
+      handler: (req, reply) => {
+        req.log.info('Auth free route')
+        reply.send({ hello: 'world' })
+      }
+    })
+
+    fastify.route({
+      method: 'GET',
+      url: '/auth',
+      beforeHandler: fastify.auth([fastify.verifyJWTandLevel.bind(fastify)]),
```

Comment:
we can avoid it either by doing  `bind` within `fastify.auth`, or by using `func.call` [here](https://github.com/fastify/fastify-auth/pull/2/files#diff-07eb183747dcb27f01bb7310efb53fc8R35)